### ISSUE TITLE
Support/standardize semantic linking of input parameters with their output objects

### ISSUE THREAD

**Issue by tycho01** [OAI/OpenAPI-Specification#728]
PR [654](https://github.com/OAI/OpenAPI-Specification/pull/654) enabled using `schema` (and therefore `$ref`) in non-body parameters, allowing one to `$ref`erence definitions to allow sharing them between input parameters and existing definitions for response objects, preventing duplication, as suggested in [#301](https://github.com/OAI/OpenAPI-Specification/issues/301) by @MrTrick and @jasonh-n-austin.

Some ways to do that:
* @MrTrick in #301 suggested `"schema": { "$ref": "#/definitions/person_id" }`, with the `Person` definition's `id` property also `$ref`ing that.
* idea 2: `"schema": { "$ref": "#/definitions/Person/properties/id" }`. Clear, but verbose.
* 3: have the `person_id` definition itself be a reference to `#/definitions/Person/properties/id` so it could function as a shorthand to `$ref` to.

I think this in itself is great, and should be encouraged to those upgrading.
And I think especially the clarity of idea 2 would enable an interesting extension: allowing one to semantically link input parameters to such object fields, to allow tools to recognize API endpoint input/output properties as related.

So I'd like to discuss how tooling vendors could reliably (without standardization of `$ref` usage) find semantic links between:
* references to definitions in response objects and the definitions they refer to ("this response includes a reference to the `User` definition")
* references to definition (fields) in input parameters, and which field of which definition they refer to ("this input parameter would be the `id` field of a `User` instance")

... with the intent of using the above information so as to suggest input information of one API endpoint could be provided by the output of another.

If one considers `$ref` purely as a pointer to be resolved, then the above ideas (even 2) contain one piece of information: anything contained at the path it resolves to.
However, this proposed semantic linking essentially requires a bit more information. Note that after resolving `$ref`s, this information goes lost:
* "the input parameter correlates to a property of the object at `#/definitions/Person`"
* "the input parameter to the property of that structure is at its sub-path `./properties/id`"
* ideally: "the structure it is part of should probably be referred to as `Person`". Note Swagger Editor solves this by [injecting](https://github.com/swagger-api/swagger-editor/blob/3e3d7d1d8210dfa2765f4996e1293b37dbd8078f/scripts/services/builder.js#L37-L47) these titles into the definitions using their keys, which seems like a fair approach.

I'd then wonder if Swagger Editor had solved point 1 as well, so as to relate response objects to their used definitions, which would obviously be similar. From what I can see though, it didn't. (Swagger Editor [shows](http://editor.swagger.io/#/) `Product` and its fields in the Uber `/products` response, but this is a resolved link, rather than also still being aware of the `Product` definition it took the fields from.)

Thoughts?

**Comment by tycho01** [comment]
Below are my reservations against altering the spec, making me lean toward non-action. The reason I'm still posting this thread is because I'm curious how others would feel about this use-case of the spec, and what might be the nicest way of pulling this off.

...

Since the same reasoning above might go for response objects, it wouldn't feel very right to change the spec, so perhaps expecting tooling authors to be smart about `$ref` resolving would be nicer -- only few people would be burdened.

I find it harder to think of a reliable way to infer the right info in the scenario of idea 1 by @MrTrick though, which is fine to the spec, and for all I know just could catch on -- it's hard to control the API specifications out there in the wild if not by standardization in the spec. And perhaps the simplest response here would be, idea 1 is semantically different, so leave it.

So I guess my question is, would anyone have ideas on how one could go about extracting these links ((1) definitions referenced in response objects, and (2) definitions or fields thereof referenced in input parameters) while we cannot enforce good practice of `$ref`s used?

**Comment by MrTrick** [comment]
It's going to be difficult.

1. The underlying json-schema concept for `$ref` does not restrict where it can point.  
   (relative vs absolute vs external as a separate issue)
2. There's no technical reason restricting where fields, parameters, responses, etc are defined. __Convention__ dictates that shared response definitions are in the `#/responses/` area, but one could just as easily (and a bit confusingly) define a response inline in one of the `#/paths/...` and `$ref` it from another path.
3. The `$ref` address has no intrinsic meaning right now. I don't think tools do anything with it other than de-referencing, and *maybe* a little bit of validation to restrict where in the spec it may point.

So, semantic linking is going to be an uphill battle!

Also, convention is usually a better approach than restriction. Many many bug tickets and discussions start with "I have a reason to do this, but the spec won't let me!" 

One more issue with an implementation guessing how parameters link to response attributes, is the plenty of cases where it's not clear;

* A parameter has to be a person id. Is it the primary key, the supervisor id, or the creating user's id?
* A parameter has to be a date. Is it related to created/modified/etc, or something else entirely?
* A parameter has to be a date. Is it an exact filter param, or a 'earlier than', or a 'later than' filter?
* Cases where an input parameter relates to mulitple response attributes.
* Cases where an input parameter relates to a different response attribute, depending.
* More complex cases that any simple spec will be unable to capture.

You'll need more than `$ref`. At the very least, response attrs will probably need to be linked somehow to filter parameters, not the other way round.

**Comment by tycho01** [comment]
I'm cool with settling for convention where alternatives fail. In that event I guess OpenAPI could still add value there by encouraging good conventions.

On ambiguities:
> A parameter has to be a person id. Is it the primary key, the supervisor id, or the creating user's id?

* I suppose secondary keys in definitions could similarly be defined + semantically linked by a `$ref` (or whatever) to the original location/definition of the primary key. If we'd be able to link both this secondary key as well as the parameter back to the original primary key, I think that should give sufficient information to allow suggesting either `id` or `supervisor_id` fields of previously obtained `User` instances (including perhaps the creating `User` you mentioned) as potentially relevant inputs for this parameter. We can't read minds, but suggesting relevant options seems like a good start.

> A parameter has to be a date. Is it related to created/modified/etc, or something else entirely?

* I think this example is illuminating here, as it shows a weakness with the `$ref` approach for this purpose. Even if created and modified dates could both share one generic date-like validation schema, having a semantic pointer resolve to that may well defeat the additional info that might have remained available had the pointers retained this distinction (despite the existing `$ref` pointing both to a common validation schema). This could form an argument in favor of extending the spec with some alternative.

> A parameter has to be a date. Is it an exact filter param, or a 'earlier than', or a 'later than' filter?

* I'll concede additional info could well go lost as you're showing here. That said, I'd argue it's still better to be able to make whatever suggestions can be drawn from the info available.

> Cases where an input parameter relates to multiple response attributes.

* I'm trying to imagine cases in additional to those mentioned above. I'm coming up with a few:
  * In the date scenario, wanting to refer to any date rather than abide by the above distinction. I suppose for the sake of convenience (no duplication) one could just `$ref` the generic date definition without any implied semantic link to certain attributes. At this point, we'd be left having to make suggestions without semantic links to certain attributes. Failing those, it should be possible to scan across the definitions for any occurrences of (sub-)schemas that would be identical to, or at least as strict as, the validation schema given for the input in question. This could include other irrelevant ones as well, but would be as good as it's gonna get without further info.
  * Say, we want a user id, but somehow the Schema (only?) defines a few user sub-classes, e.g. buyers and sellers. If a `User` base class is also defined, then at that point it'd seem to come down to whether the 'sub-classes' would reuse that in a way that could be traced back. This'd return to the original question of tracing back inclusions of definitions in other definitions.
  * In the above case, but without a reused base class, it'd get more complicated. If this is a valid use-case, it may be desirable to have these 'links' be an array, so as to take multiple such options into account. I cannot think of good cases, and would think they may well point to bad design, but yeah, for all I know it could happen.

> Cases where an input parameter relates to a different response attribute, depending.

* I guess this'd be similar to the last mentioned scenario. Yeah, perhaps these should be taken into account as well.

> More complex cases that any simple spec will be unable to capture.

I hope the above may help a bit here, and agree there might be logic that'd be beyond the spec. I guess overall though, I'm okay with not having something perfect, as I think API descriptions would already become more valuable with the ability to make some basic inferences of what obtained info could be plugged into what other endpoints.

> You'll need more than `$ref`. At the very least, response attrs will probably need to be linked somehow to filter parameters, not the other way round.

This differs from my conclusion, so might I ask you to elaborate on this further?
My reasoning: there may well be many API endpoints taking a user id as an input parameter. Admittedly, the number of links might be similar either way... If I were writing an API description though, it would seem easier if I could add an endpoint by being able to copy-paste parts from an existing one, rather than having to add to an array halfway across the document as well. (Similarly, when writing ES6, I might prefer to use `export function` rather than keeping track of a separate list of anything to be exported at the bottom of the file.)

If `$ref` would indeed not suffice here though, as it seems may well be the case, I hope others might see the potential added value here as well.

**Comment by MrTrick** [comment]
>> You'll need more than `$ref`. At the very least, response attrs will probably need to be linked somehow to filter parameters, not the other way round.
>
>This differs from my conclusion, so might I ask you to elaborate on this further?

* That `$ref` is an existing construct already used in all kinds of ways. Overloading and restricting it by way of adding semantic meaning is likely to make life harder for spec designers.
* Abstracting for a moment the specific keyword used, there are two ways to have that link in the spec; "in param define effect on response" or "in response define how param affects it". I'll amend my previous comment, and say that either are doable.

I do applaud your idea, but I'm not exactly sure whether you are trying to define a filtering spec, or define a mechanism by which a filtering spec may itself be defined? 

(If the former, then remember that the OpenAPI spec is meant to be API agnostic)  
(If the latter, then consider just how difficult it will be to support any API design!)

**Comment by tycho01** [comment]
Sorry, yeah, I should've bolded the second sentence in your quote, the first half was clear. I think I agree with you on `$ref`.

> I'm not exactly sure whether you are trying to define a filtering spec, or define a mechanism by which a filtering spec may itself be defined?

I'm not sure I'm fully grasping the distinction here. Maybe I'd be better off reconfirming my understanding first.
I'm thinking in terms of 'relevant suggestions' rather than 'filter' though. This difference again sounds subtle, but the latter sounds like adding limitations, while I'd think of having better information as enabling rather than limiting.
I *think* that'd put me in the latter category, if I haven't misunderstood your intent.

You're probably right others may well find room for improvement. In fact, I hope so -- I can't foresee everything by myself.
Would it be counter-productive to approve a proposal without sufficient outside input? Yeah, probably.

Maybe I'd get a bit further by attempting to flesh something out so as to find the obstacles along the way. So let me try if I can flesh out something more concrete, as it'll probably be easier to criticize/improve.

**Comment by tycho01** [comment]
Optional point:
* Does it matter if all 'model' definitions could be found in the same place (e.g. `#/definitions`?). From a tooling perspective, predictability is nice, and Swagger Editor seems to manage to list any 'models' defined over there, which seems useful. From a definition writer perspective, for all I know some will find it desirable to place them elsewhere, e.g. even in external files (presuming the spec allows this). Pushing for standardization there seems counter-productive, while this linking proposal may also facilitate the discovery of models. There may be additional ways to facilitate discoverability of models as well. One idea would be to standardize a `models` property in the spec, as an array of locations (e.g. `["#/definitions/Person"]`), as an object of such locations if the names cannot or should not be extracted from those paths (e.g. `{ "Person": "#/definitions/Person" }`) for whatever reason (?), or perhaps a `modelPaths` array of locations where such models can be found (e.g. `["#/definitions"]`).

On `$ref` alternatives:
* So one question was if `#/definitions/Person/properties/id` would contain sufficient info to infer (a) we're looking at a field of this model at `#/definitions/Person`, (b) that model is called `Person`, and (3) we're looking for its field at `properties/id`. I think one can answer all of those with 'yes' presuming (I) `json-schema` keywords like `items`, `additionalItems`, `properties`, `patternProperties`, `additionalProperties`, `allOf`, `anyOf`, `oneOf` and similar can be seen as indications that we've started navigating a model, and (II) we can just take the last path slice of the model (`Person`) as its name. If these assumptions appear to be somewhat sane/reasonable, then I think syntax would need to be no more complex than that of `$ref`.
* As a potential new keyword, `link` or whatnot works for all I'm concerned, but I'd have settled for `$softref` with original syntax intact except expecting a direct link (the "idea 2" above -- for all I care if people wanted to make a `person_id` shorthand, as in idea 3 in the OP, they could have the `$softref` placed in `#/definitions/person_id`, so they could use regular `$ref`s to that so as to use it as a shorthand alias). You noted that making it an array may preferable though, in which case the keyword name should probably reflect that it's not identical to `$ref`. I don't know -- `links`? That being said, I suppose it would have been nice if changes would be minimal for those who do not care about this extra information, but would rather just use them as traditional `$ref`s. But I suppose, array considerations aside, the keyword cannot both be the same and different. Unless...
* Alternatively, I suppose we could allow a boolean called e.g. `softLink` to be used in conjunction with `$ref`, e.g. `{ "$ref": "#/definitions/Person/properties/id", "softLink": true }` (with the location expected to be a direct reference for reasons explained two points above in a/b/c). This would minimize burden on tool-writers and spec-writers not interested in the additional functionality by retaining backward compatibility, while providing potentially sufficient information for the purpose of semantic linking under assumptions I/II. Meanwhile, tool writers interested in semantic linking could opt to only directly resolve `$ref`s not marked soft, take the desired info from the remaining links, and then, I suppose, resolve those as well.
* This last idea is again ignoring your earlier comment on allow to reference multiple items though. I'd wonder if using `anyOf` in combination with something like this `softLink` might do there...

**Comment by MrTrick** [comment]
@tycho01 I think the more fundamental issue is deeper than any `$ref` or specific keywords.

Consider the docs for the OpenAPI spec itself; "OpenAPI does not require you to rewrite your existing API. It does not require binding any software to a service--the service being described may not even be yours. It does, however, require the capabilities of the service be described in the structure of the OpenAPI Specification."  
So let's take that as being "you should be able to implement a sensible API, (I know the docs go on to say it won't cover *all* APIs) and have the spec describe it".

"Linking input parameters with their output objects" sounds to me like potentially a large flexible space. API A might use input parameters in a completely different manner to API B.  
Consider carefully what you'd like to achieve in an extension or enhancement here:

* **Option 1:** Define a mechanism that improves the setup, code generation etc process for "API A"-like APIs. Hooray! - **unless you have an API that's not "API A"-like.**
* **Option 2:** Define a mechanism that tries to improve the setup, code generation etc process for any kind of API - A, B, C... Much more difficult to do, and the creator of API D might still be unsatisfied with the options available within the spec.

This sounds to me like a miniature version of what I've been cross with Swagger/OpenAPI about in the last few months; 
 1. The "JSON-API" specification is one sensible way to design/implement an API.
 2. I've built a system that meets the spec, and I want to define a corresponding swagger spec, for documentation, etc.
 3. The Swagger/OpenAPI standard made it really difficult! 
      * The existing json-schema describing JSON-API was considered invalid by strict swagger standards.
      * Many of the JSON-API standards weren't describable inside the swagger spec.
      * Much of parameters' semantic meaning (like `$ref:"#/definitions/person_id") was lost.
 4. So I'm stuck with keeping two spec documents; 
      * one that's got all the features I want around validation, sanity etc. and 
      * one that's dumbed down, but that conforms to the OpenAPI spec.

Moral of the story? Don't go with **Option 2** - that way lies disappointment confusion and anger. If not for the one defining that lovely mechanism, then for every author that comes around and is unable to use it to capture how their API works.

**Comment by tycho01** [comment]
Thanks, that's a fair point. I'm definitely okay with not covering every single edge case.
And perhaps by taking a few assumptions (notably (I), I'm more open about (II)) I'm already trying to compromise a few edge cases for the sake of more common cases (tooling/description users that don't care and would rather have higher backward compatibility).

**Comment by tycho01** [comment]
I tried a quick [PoC](https://gist.github.com/tycho01/179c6c0dbfe6d06681ee01314892e799) (i.e. I glossed over the complexity of JSON reference/pointer e.g. external/remote files) to demonstrate it's possible to gather the info with just `$ref`, so no changes to the OpenAPI spec, though @MrTrick's scenario ends up a bit different.
As stated above, one can't just run the existing de`$ref`erencer in your language over an api spec and expect that to get you this info, but I'm happy this seems viable without changes.
